libraries/AP_WayBack/AP_WayBack.cpp

```diff
@@ -0,0 +1,818 @@
+    simplify(last_reduce, num_points); // reduce all remaining points
+
+    // remove last loops
+    uint16_t p1=points_count-1; // last good point
+    
+    while(last_loop_check > p1){
+        uint16_t p0 = move_back(p1,1); // points can be not removed so we use relative coordinates
+         // check segment for track intersection
+
+    DBG_PRINT("check loop1 ");     DBG_PRINTVAR(p0);    DBG_PRINTVARLN(p1);
+
+        uint16_t sb = try_remove_loop(p0, p1);
+
+        if(sb) { // loop found
+            p0=move_back(sb,1);    // prev point 
+    DBG_PRINT("loop found at ");       DBG_PRINTVARLN(p1);            
+
+            simplify(p0, p1); // try to reduce after removal
+            break; // we move back so just find shortest path
+        }
+        p1 = p0; // set left point as right    
+    }
+
+}
+
+// get last point and remove it from track
+bool AP_WayBack::get_point(float &x, float &y) {
+//    if(recording) stop(); Is we need it?
+    
+    for(uint16_t i=num_points;i>0;i--){ 
+        uint16_t idx = i-1;
+        if(is_good(points[idx])){ // found a good point
+            num_points=idx; // remove last
+            
+            x=points[idx].x;
+            y=points[idx].y;
+            return true;
+        }
+    }
+    return false; // no more points so go directly to HOME point
+}
+
+/*
+ * tick - main call to get data from AHRS
+ */
+void AP_WayBack::tick(void)
+{ 
+
+    if(points==NULL) return;
+
+    uint32_t now = AP_HAL::millis();
+    uint32_t dt  = now - last_point_time;
+    if( dt < 1000) {    // 1 point per second
+        yield(1000-dt); // don't disturbe for needed time
+        return; 
+    }
+
+#if defined(WAYBACK_DEBUG)
+    if(_debug_mode) return; // work in soft emulation
+#endif
+
+    if(recording){
+
+// here we should to get a current UAV location filtered with EKF to prevent recording of GPS glitches
+#if FRAME_CONFIG ==  MULTICOPTER_FRAME
+        Vector3f pos;
+        _ahrs.get_relative_position_NED_home(pos);
+        add_point(pos.x, pos.y);
+
+#else
+        Location loc;
+        if (!_ahrs.get_position(loc)) {
+            return;
+        }
+        add_point(loc.lat, loc.lng);
+#endif
+    }
+
+}
+
+
+
+#if defined(WAYBACK_DEBUG)
+// print out track points
+bool AP_WayBack::show_track(uint16_t &i, float &x, float &y ) {
+    while(i<num_points){ 
+        if(is_good(points[i])){ // found a good point
+            x=points[i].x;
+            y=points[i].y;
+            i++;
+            return true;
+        }
+        i++;
+    }
+    return false;
+}
+#endif
+
+
+
+// private 
+
+// move back on STEPS good points
+uint16_t AP_WayBack::move_back(uint16_t from, uint16_t steps){
+    if(from==0) return 0;
+    for(uint16_t i=from-1;i>0;i--){ 
+        if(is_good(points[i])){ // found a good point
+            if(--steps == 0) return i;
+        }
+    }
+
+    return 0; //failed
+}
+
+
+// move forward on STEPS good points
+uint16_t AP_WayBack::move_forw(uint16_t from, uint16_t steps){
```

Comment:
writing code for speed is good, but if its hard to read and is a different style than the rest of the codebase then it becomes unmaintainable and that's far worse to have in your codebase. Therefore, we have set up some guidelines. Nothing personal.